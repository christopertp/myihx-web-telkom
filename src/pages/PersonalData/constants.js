export const FAILED = 'PersonalData/failed';
export const FETCHED = 'PersonalData/fetched';
export const LOADING = 'PersonalData/loading';
