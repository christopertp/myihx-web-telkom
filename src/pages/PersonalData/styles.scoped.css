.review-image {
  border-radius: 0.5rem;
  margin-bottom: 1rem;
  width: 100%;
}

.error {
  color: #ee4c24;
  display: block;
  font-size: 0.75rem;
  margin-top: 1rem;
}

.reupload {
  display: flex;
  justify-content: space-between;
  margin-top: 2rem;
}

.reupload > * {
  flex: 1
}

.reupload > *:first-child {
  margin-right: 1rem;
}

.verify-button {
  margin-top: 2rem;  
}

@media only screen and (max-width: 48em) {
  .verify-button {
    margin: auto 0 1.5rem;
  }
}
