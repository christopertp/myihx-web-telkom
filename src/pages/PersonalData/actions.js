import { push } from 'react-router-redux';
import { document, documentConfirm, documentUpdate, documentUpload, documentUploadSelfi, documentUploadSignature } from '../../services/user';
import { FAILED, FETCHED, LOADING } from './constants';

export function fetchConfirm() {
  return async dispatch => {
    dispatch(loadingAction(true));
    try {
      await documentConfirm();
      dispatch(failedAction(''));
      dispatch(push('/personal-data/verify'));
    } catch (err) {
      dispatch(failedAction(err.message));
    }
  };
}

export function fetchDocument() {
  return async dispatch => {
    dispatch(loadingAction(true));
    try {
      const { data } = await document();
      dispatch(fetchedAction(data));
    } catch (err) {
      dispatch(fetchedAction(''));
    }
  };
}

export function fetchUpload(type, payload, subpage) {
  const fn = {
    id: subpage === 'edit' ? documentUpdate : documentUpload,
    selfie: documentUploadSelfi,
    sign: documentUploadSignature,
  };
  return async dispatch => {
    dispatch(loadingAction(true));
    try {
      await fn[type](payload);
      dispatch(failedAction(''));
      dispatch(push('/personal-data'));
    } catch (err) {
      dispatch(failedAction(err.message));
    }
  };
}

function failedAction(message) {
  return { type: FAILED, message };
}

function fetchedAction(data) {
  return { type: FETCHED, data };
}

function loadingAction(isLoading) {
  return { type: LOADING, isLoading };
}
