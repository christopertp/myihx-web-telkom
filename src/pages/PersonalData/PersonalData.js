import React, { createContext, useEffect, useState, useContext, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory, useParams } from 'react-router-dom';
import Button from '../../components/elements/Button';
import Spinner from '../../components/elements/Spinner';
import Summary from '../../components/fragments/PersonalDataSummary';
import ReviewForm from '../../components/forms/PersonalDataReview';
import WhiteCardPlain from '../../components/layouts/WhiteCardPlain';
import WhiteCardSuccessPlain from '../../components/layouts/WhiteCardSuccessPlain';
import { fetchDocument, fetchUpload } from './actions';
import styles from './styles.scoped.css';

const Context = createContext({});

export default function PersonalData() {
  const history = useHistory();
  const { page } = useParams();
  const [photo, setPhoto] = useState('');

  useEffect(() => {
    (page && page !== 'verify') && history.replace('/personal-data');
  }, []);

  const contextValue = {
    photo, setPhoto,
  };

  return (
    <Context.Provider value={contextValue}>
      {(p => {
        if (['e-ktp', 'ktp', 'passport', 'sim'].includes(p)) {
          return <ReviewId />;
        }
        if (['selfie', 'sign'].includes(p)) {
          return <ReviewSelfieSign />;
        }
        if (p === 'verify') {
          return <Verify />;
        }
        return <Summary fetchDocument={fetchDocument} setPhoto={setPhoto} />;
      })(page)}
    </Context.Provider>
  );
}

export function ReviewId() {
  const { photo, setPhoto } = useContext(Context);
  const dispatch = useDispatch();
  const { page, subpage } = useParams();
  const { onChangeFile, onRetake, preview, refFile } = usePhoto(photo, setPhoto);
  const { photos } = useSelector(s => s.personalData);
  const subtitle = 'Please make sure your picture is clear and show complete information.';

  const onSubmit = v => {
    const payload = new FormData();
    Object.keys(v).forEach(i => {
      payload.append(i, v[i]);
    });
    payload.append('type', page.replace('-',''));
    payload.append('image', photo);
    dispatch(fetchUpload('id', payload, subpage));
  };

  useEffect(() => () => setPhoto(null), []);

  return (
    <WhiteCardPlain back="/personal-data" subtitle={subtitle} title="Review Photo">
      <img alt="" className={styles['review-image']} src={preview ? preview : photos.id} />
      <ReviewForm onRetake={onRetake} onSubmit={onSubmit} />
      <input accept="image/png, image/jpeg, image/jpg" onChange={onChangeFile} ref={refFile} style={{ display: 'none' }} type="file" />
    </WhiteCardPlain>
  );
}

export function ReviewSelfieSign() {
  const { photo, setPhoto } = useContext(Context);
  const dispatch = useDispatch();
  const { page } = useParams();
  const { onChangeFile, onRetake, preview, refFile } = usePhoto(photo, setPhoto);
  const { isLoading, message } = useSelector(s => s.personalData);
  const title = page === 'selfie' ? 'Review Photo' : 'Review Signature Image';
  const subtitle = page === 'selfie' ?
    'Please make sure your picture is clear and show complete information.' :
    'Please make sure your signature picture is clear.';

  const onSubmit = () => {
    const payload = new FormData();
    payload.append('image', photo);
    dispatch(fetchUpload(page, payload));
  };

  return (
    <WhiteCardPlain back="/personal-data" subtitle={subtitle} title={title}>
      <img alt="" className={styles['review-image']} src={preview} />
      {message && <small className={styles.error}>{message}</small>}
      <div className={styles.reupload}>
        <Button disabled={isLoading} onClick={onRetake} variant="bordered">Reupload</Button>
        <Button disabled={isLoading} onClick={onSubmit}>
          {isLoading ? <Spinner /> : 'Confirm'}
        </Button>
      </div>
      <input accept="image/png, image/jpeg, image/jpg" onChange={onChangeFile} ref={refFile} style={{ display: 'none' }} type="file" />
    </WhiteCardPlain>
  );
}

export function Verify() {
  const title = 'Verification in Progress';
  const subtitle = 'Please wait while weâ€™re verifying your ID. You will receive a phone call from our agent within 1x24 hours.';
  return (
    <WhiteCardSuccessPlain icon="waiting" subtitle={subtitle} title={title}>
      <Button className={styles['verify-button']} to="/">Back to Home</Button>
    </WhiteCardSuccessPlain>
  );
}

export function usePhoto(photo, setPhoto) {
  const refFile = useRef(null);
  const [preview, setPreview] = useState('');

  useEffect(() => {
    const reader = new FileReader();
    reader.onload = e => {
      setPreview(e.target.result);
    };
    photo && reader.readAsDataURL(photo);
  }, [photo]);

  const onRetake = () => {
    refFile.current.click();
  };

  const onChangeFile = e => {
    const { files } = e.target;
    setPhoto(files[0]);
  };

  return { onChangeFile, onRetake, preview, refFile };
}
