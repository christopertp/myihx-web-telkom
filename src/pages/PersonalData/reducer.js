import { fetchConfirm } from './actions';
import { FAILED, FETCHED, LOADING } from './constants';

const initialState = {
  fetchConfirm,
  invalid: '',
  isLoading: true,
  message: '',
  form: {
    dateOfBirth: '',
    idNumber: '',
    motherName: '',
    type: '',
  },
  photos: {
    id: '',
    selfie: '',
    sign: '',
  },
  timeInSeconds: (24 * 60 * 60) - 1,
};

export default function reducer(state = initialState, action = {}) {
  const { type, data, isLoading, message } = action;

  switch (type) {
    case FAILED:
      return {
        ...state,
        isLoading: false,
        message,
      };
    case FETCHED:
      return {
        ...state,
        form: data ? {
          dateOfBirth: data.dateOfBirth,
          idNumber: data.idNumber,
          motherName: data.motherName,
          type: data.type,
        } : '',
        photos: data ? {
          id: data.identity,
          selfie: data.selfiWithId,
          sign: data.signature,
        } : '',
        isLoading: false,
        invalid: (data && data.validation) ? data.validation.invalidReason : '',
        message: '',
        timeInSeconds: data
          ? (initialState.timeInSeconds - data.timeInSeconds)
          : state.timeInSeconds,
      };
    case LOADING:
      return {
        ...state,
        isLoading,
      };
    default:
      return state;
  }
}
