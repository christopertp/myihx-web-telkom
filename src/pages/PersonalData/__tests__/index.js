import React, { useEffect, useContext, useState, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory, useParams } from 'react-router-dom';
import ShallowRenderer from 'react-test-renderer/shallow';
import Spinner from '../../../components/elements/Spinner';
import PersonalData, { ReviewId, ReviewSelfieSign, Verify, usePhoto } from '../PersonalData';

jest.mock('../actions', () => ({
  fetchDocument: jest.fn(),
  fetchUpload: jest.fn(),
}));

useSelector.mockImplementation(fn => {
  fn({});
  return { photos: '' };
});

describe('src/pages/PersonalData', () => {
  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<PersonalData />);
    expect(tree).toMatchSnapshot();
  });

  test('render subbroutes', () => {
    const replace = jest.fn();
    useHistory.mockImplementationOnce(() => ({ replace }));

    useParams.mockImplementationOnce(() => ({ page: 'sim' }));
    const result1 = PersonalData();
    expect(result1.props.children.type).toBe(ReviewId);
    expect(replace).toHaveBeenCalledWith('/personal-data');

    useParams.mockImplementationOnce(() => ({ page: 'sign' }));
    const result2 = PersonalData();
    expect(result2.props.children.type).toBe(ReviewSelfieSign);

    useParams.mockImplementationOnce(() => ({ page: 'verify' }));
    const result3 = PersonalData();
    expect(result3.props.children.type).toBe(Verify);
  });

  test('ReviewId', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<ReviewId />);
    expect(tree).toMatchSnapshot();
  });

  test('ReviewId onSubmit & onUnmout', () => {
    const dispatch = jest.fn();
    useDispatch.mockImplementationOnce(() => dispatch);

    const setPhoto = jest.fn();
    useState.mockImplementationOnce(() => ['preview', jest.fn()]);
    useContext.mockImplementationOnce(() => ({ setPhoto }));

    useEffect.mockImplementationOnce(fn => fn()).mockImplementationOnce(fn => fn()());
    useParams.mockImplementationOnce(() => ({ page: '' }));

    const result = ReviewId();
    result.props.children[1].props.onSubmit({ test: 'test' });
    expect(dispatch).toHaveBeenCalled();
    expect(setPhoto).toHaveBeenCalledWith(null);
  });

  test('ReviewSelfieSign', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<ReviewSelfieSign />);
    expect(tree).toMatchSnapshot();
  });

  test('ReviewSelfieSign selfie page, loading/error state & onSubmit', () => {
    const dispatch = jest.fn();
    useDispatch.mockImplementationOnce(() => dispatch);

    useParams.mockImplementationOnce(() => ({ page: 'selfie' }));
    useSelector.mockImplementationOnce(() => ({ isLoading: true, message: 'error' }));

    const result = ReviewSelfieSign();
    expect(result.props.title).toBe('Review Photo');
    expect(result.props.children[1].props.children).toBe('error');

    const button = result.props.children[2].props.children[1];
    expect(button.props.children.type).toBe(Spinner);
    button.props.onClick();
    expect(dispatch).toHaveBeenCalled();
  });

  test('Verify', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Verify />);
    expect(tree).toMatchSnapshot();
  });

  test('usePhoto', () => {
    const click = jest.fn();
    useRef.mockImplementationOnce(() => ({ current: { click } }));

    const setPreview = jest.fn();
    useState.mockImplementationOnce(() => ['preview', setPreview]);

    delete global.FileReader;
    const readAsDataURL = jest.fn();
    const reader = { readAsDataURL };
    global.FileReader = () => reader;

    const setPhoto = jest.fn();
    const { onChangeFile, onRetake } = usePhoto('photo', setPhoto);

    expect(readAsDataURL).toHaveBeenCalledWith('photo');

    reader.onload({ target: { result: 'res' } });
    expect(setPreview).toHaveBeenCalledWith('res');

    onRetake();
    expect(click).toHaveBeenCalled();

    const files = ['file'];
    onChangeFile({ target: { files } });
    expect(setPhoto).toHaveBeenCalledWith('file');
  });
});
