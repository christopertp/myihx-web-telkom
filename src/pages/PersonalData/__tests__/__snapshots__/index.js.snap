// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/pages/PersonalData ReviewId 1`] = `
<WhiteCardPlain
  back="/personal-data"
  close=""
  help={false}
  icon=""
  size="large"
  subtitle="Please make sure your picture is clear and show complete information."
  title="Review Photo"
>
  <img
    alt=""
  />
  <ReduxForm
    destroyOnUnmount={true}
    enableReinitialize={false}
    forceUnregisterOnUnmount={false}
    form="personalDataReview"
    getFormState={[Function]}
    initialValues={
      Object {
        "idNumber": "",
        "motherName": "",
        "name": "",
      }
    }
    keepDirtyOnReinitialize={false}
    onRetake={[Function]}
    onSubmit={[Function]}
    persistentSubmitErrors={false}
    pure={true}
    shouldAsyncValidate={[Function]}
    shouldError={[Function]}
    shouldValidate={[Function]}
    shouldWarn={[Function]}
    submitAsSideEffect={false}
    touchOnBlur={true}
    touchOnChange={false}
    updateUnregisteredFields={false}
    validate={[Function]}
  />
  <input
    accept="image/png, image/jpeg, image/jpg"
    onChange={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    type="file"
  />
</WhiteCardPlain>
`;

exports[`src/pages/PersonalData ReviewSelfieSign 1`] = `
<WhiteCardPlain
  back="/personal-data"
  close=""
  help={false}
  icon=""
  size="large"
  subtitle="Please make sure your signature picture is clear."
  title="Review Signature Image"
>
  <img
    alt=""
    src=""
  />
  <div>
    <Button
      className=""
      disabled={false}
      onClick={[Function]}
      to=""
      type="button"
      variant="bordered"
    >
      Reupload
    </Button>
    <Button
      className=""
      disabled={false}
      onClick={[Function]}
      to=""
      type="button"
      variant="red"
    >
      Confirm
    </Button>
  </div>
  <input
    accept="image/png, image/jpeg, image/jpg"
    onChange={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    type="file"
  />
</WhiteCardPlain>
`;

exports[`src/pages/PersonalData Verify 1`] = `
<WhiteCardSuccessPlain
  close=""
  icon="waiting"
  subtitle="Please wait while weâ€™re verifying your ID. You will receive a phone call from our agent within 1x24 hours."
  title="Verification in Progress"
>
  <Button
    className=""
    disabled={false}
    onClick={null}
    to="/"
    type="button"
    variant="red"
  >
    Back to Home
  </Button>
</WhiteCardSuccessPlain>
`;

exports[`src/pages/PersonalData render 1`] = `
<Provider
  value={
    Object {
      "photo": "",
      "setPhoto": [MockFunction],
    }
  }
>
  <PersonalDataSummary
    fetchDocument={[MockFunction]}
    setPhoto={[MockFunction]}
  />
</Provider>
`;
